### PR TITLE
[NEW] PDF message attachment preview (client side rendering)

### LINKED ISSUE #2741: Make previews for PDF files
Your Rocket.Chat version: 0.24.0
It would be great to make previews for first page of PDF-book, attached or linked to the chat.

### PR DESCRIPTION
**Pull request by kb0304**
Preview of PDF files
Rendering first page on a canvas using a `Worker` on client side

Closes #2741
pdf-preview

### CODE REVIEW

**Comment by kb0304** [comment]
@graywolf336 Sure, here's the  [link](https://github.com/kb0304/meteor-pdfjs)

**Comment by kb0304** [comment]
@graywolf336 
Its just a wrapper for Mozilla's pdf.js ([source code](https://github.com/mozilla/pdf.js))
The build files are taken from [pdfjs-dist](https://github.com/mozilla/pdfjs-dist), and just added [this](https://github.com/kb0304/meteor-pdfjs/blob/master/build/pdf.js#L34) to export globaly.

I initially used [pascoual:pdfjs](https://github.com/pascoual/meteor-pdfjs), but since the version of pdf.js used is outdated and it no longer seems to be maintained I made a wrapper myself and used it.

Documentations, examples to use it are available [here](https://mozilla.github.io/pdf.js/examples/)

**Comment by tobzsc** [comment]
Any news here? That's a very nice feature.

File: packages/rocketchat-ui-message/client/message.js [hunk 1]
```diff
@@ -1,7 +1,31 @@
-/* globals renderEmoji renderMessageBody */
+/* globals pdfjsLib renderEmoji renderMessageBody */
 import _ from 'underscore';
 import moment from 'moment';
 
+pdfjsLib.GlobalWorkerOptions.workerSrc = '/packages/kb0304_pdfjs/build/pdf.worker.js';
```
Comment (tassoevan) [inline packages/rocketchat-ui-message/client/message.js]: This path can cause some troubles if Rocket.Chat is running under a root URL with a subdirectory.

File: packages/rocketchat-ui-message/client/message.js [hunk 2]
```diff
@@ -357,7 +381,10 @@ Template.message.onCreated(function() {
 });
 
 Template.message.onViewRendered = function(context) {
-	return this._domrange.onAttached(function(domRange) {
+	return this._domrange.onAttached((domRange) => {
```
Comment (tassoevan) [inline packages/rocketchat-ui-message/client/message.js]: @ggazzo Do you know if Meteor provides a public method to access this property?

**Comment by nopynoop** [comment]
Any news on this? Looks like a very helpful addition to the Rocket.Chat!

**Comment by rafaelks** [comment]
This pull-request started crashing Safari 12, I cannot share the GIF here because there's sensitive information on it, but I shared with the internal team already.

**Comment by imeleges** [comment]
Does this feature needs to be enabled somewhere in the administration settings?
As it doesn't show preview of any pdf sent to a direct message or group. I have SentOS 7 and version 0.71.0 of RC

**Comment by tobzsc** [comment]
I have the same issue. Does not seem to work with latest stable.

**Comment by beranPro** [comment]
https://github.com/RocketChat/Rocket.Chat/issues/12449

**Comment by reetp** [comment]
Tested on my own 0.71.1 test setup and on https://openrocket.chat/general
No apparent errors on upload - file is there and can be downloaded, but no preview/thumbnail

**Comment by flam22** [comment]
The reason the pdf preview no longer works is that it was broken by PR #12375. An extra test was added to disable the preview in safari versions <= 12. That also affects Chrome (which may or may not be intentional).

In addition the original first line of 'renderPdfToCanvas'  was changed to use a regular expression but a '!' was omitted, which stops it ever producing a preview. It should read:
`if (!pdfLink || ! /\.pdf$/i.test(pdfLink)) { return; }`

**Pull request merged**